lnd_test.go

```diff
@@ -3644,6 +3645,462 @@ func testMultiHopPayments(net *lntest.NetworkHarness, t *harnessTest) {
+		succeeded := make([]uint64, numPayIntents)
+		tried := make([]uint64, numPayIntents)
+		var totSucceeded uint64 = 0
+		var totTried uint64 = 0
+		for i := 0; i < numPayIntents; i++ {
+			succeeded[i] = atomic.LoadUint64(&atomicSucceeded[i])
+			tried[i] = atomic.LoadUint64(&atomicTried[i])
+			totSucceeded += succeeded[i]
+			totTried += tried[i]
+		}
+		mux.Lock()
+		fmt.Printf("%v transactions performed in total, %v succeeded\n", totTried, totSucceeded)
+		fmt.Printf("Rate: %.2f%%\n", 100.0 * float64(totSucceeded) / float64(totTried))
+		for i := 0; i < numPayIntents; i++ {
+			rt := 100.0 * float64(succeeded[i]) / float64(tried[i])
+			fmt.Printf("%v -> %v: %.2f%%\t(%v/%v)\n", nodeNames[payIntents[i][0]], nodeNames[payIntents[i][1]], rt,
+				succeeded[i], tried[i])
+		}
+		mux.Unlock()
+	}
+
+	// Helper function to lookup node name based on channel id
+	lookupChan := func(chanID uint64) (string, string) {
+		for i := 0; i < numChannels; i++ {
+			if chanIDs[i] == chanID {
+				start := nodeNames[connections[i][0]]
+				end := nodeNames[connections[i][1]]
+				return start, end
+			}
+		}
+		return "U", "U"	// U for unknown
+	}
+
+	// Helper function to dump channel info
+	printChan := func(chanIdx int) {
+		listReq := &lnrpc.ListChannelsRequest{}
+		listResp, _ := nodes[connections[chanIdx][0]].ListChannels(ctxb, listReq)
+		localChannels := listResp.Channels
+		for _, ch := range localChannels {
+			if ch.ChanId == chanIDs[chanIdx] {
+				mux.Lock()
+				fmt.Printf("%v->%v stat: Loc. Bal. %v, Rmt. Bal. %v, Sent %v, Rcvd %v\n",
+					nodeNames[connections[chanIdx][0]], nodeNames[connections[chanIdx][1]],
+					ch.LocalBalance, ch.RemoteBalance, ch.TotalSatoshisSent, ch.TotalSatoshisReceived)
+				mux.Unlock()
+			}
+		}
+	}
+
+	printRoute := func(startIdx int, hops []*lnrpc.Hop) {
+		lastNodeName := nodeNames[startIdx]
+		for hopIdx, hop := range hops {
+			start, end := lookupChan(hop.ChanId)
+			if start == lastNodeName {
+				lastNodeName = end
+				if hopIdx == 0 {
+					fmt.Printf("%v->%v", start, end)
+				} else {
+					fmt.Printf("->%v", end)
+				}
+			} else {
+				lastNodeName = start
+				if hopIdx == 0 {
+					fmt.Printf("%v->%v", end, start)
+				} else {
+					fmt.Printf("->%v", start)
+				}
+			}
+		}
+	}
+
+	// Send payments.
+
+	// For each payment demand, we create a "dispatcher" goroutine to create transactions for that demand.
+	// Then we use channels to control when they can fire a payment. Channels are filled by another goroutine
+	// that periodically sends to those channels.
+
+	// Channels to control payment goroutines.
+	var paymentTick = make([]chan int, numPayIntents)
+	for i := range paymentTick {
+   		paymentTick[i] = make(chan int, 1)
+	}
+
+	// Channels to signal the end of the experiment
+	var paymentStop = make([]chan int, numPayIntents)
+	for i := range paymentTick {
+   		paymentStop[i] = make(chan int, 1)
+	}
+	
+	// Dispatcher goroutines
+	for i := 0; i < numPayIntents; i++ {
```

Comment:
Any chance you can refactor this code and give all the goroutines names? Its just really hard to follow the whole 3 layers of goroutines. I get the flow - but might help if there were some descriptive names/self contained functions.